pyvast/pyvast/vast.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+
+"""python vast module
+
+Disclaimer: `await` does not work in the python3.7 repl,
+use either python3.8 or ipython.
+
+Example:
+
+    Create a connector to a VAST server:
+    > from pyvast import VAST
+    > vast = VAST(binary="/opt/tenzir/bin/vast")
+    Test if the connector works:
+    > await vast.test_connection()
+    Extract some Data:
+    > data = await vast.export(max_events=10).json(":addr == 192.168.1.104").exec()
+
+"""
+
+import asyncio
+import logging
+
+
+class VAST:
+    """A VAST node handle"""
+
+    def __init__(self, binary="vast", endpoint="localhost:42000"):
+        self.logger = logging.getLogger("vast")
+        self.binary = binary
+        self.endpoint = endpoint
+        self.call_stack = []
+        self.logger.debug(f"VAST client configured to use endpoint {self.endpoint}")
+
+    async def __spawn(self, *args, stdin=None):
+        """Spawns a process asynchronously."""
+        return await asyncio.create_subprocess_exec(
+            self.binary,
+            "-e",
+            self.endpoint,
+            *args,
+            stdin=stdin,
+            stdout=asyncio.subprocess.PIPE,
+        )
+
+    async def test_connection(self):
+        """Checks if the configured endpoint is reachable"""
+        proc = await self.__spawn("status")
+        stdout, stderr = await proc.communicate()
+        return proc.returncode == 0
+
+    async def exec(self, stdin=None):
+        self.logger.debug(f"Executing call stack: {self.call_stack}")
+        if not stdin:
+            proc = await self.__spawn(*self.call_stack)
+        else:
+            self.logger.debug(f"Forwarding stdin {stdin}")
+            proc = await self.__spawn(*self.call_stack, stdin=asyncio.subprocess.PIPE)
+            proc.stdin.write(str(stdin).encode())
+            await proc.stdin.drain()
+            proc.stdin.close()
+        self.call_stack = []
+        return await proc.communicate()
+
+    def __getattr__(self, name, **kwargs):
+        """Chains every unknown method call to the internal call stack."""
+        if name.startswith("_"):
+            # trim leading underscores to overcome the 'import' keyword
+            name = name[1:]
+        self.call_stack.append(name)
```

Comment:
@tobim mentioned that this flexibility comes with the downside breaking stuff like autocompletion when working with PyVAST.

What if instead of deciding on either the dynamic or the hardcoded way, we kept the dynamic way, but additionally hardcode a few methods like `start`, `export` or `import`—the things that are done the most often.

Reply:
@tobim and I discussed yesterday that for this MVP version of the library it is OK to not have autocompletion in IDEs. We already have a follow-up story for auto-generating a complete command skeleton from the binary (ch14424). That story also covers the ability to fail early and pass error messages from the python library to the user, rather than from the vast process itself (as it is now).